File: C++/CLRS/Merge_Sort.cpp
#include <bits/stdc++.h>
using namespace std;

void Merge(int A[], int p, int q, int r)
{
    int n_l=q-p+1;
    int n_r=r-q;

    int L[n_l],R[n_r];

    int i,j;
    for (i=0;i<n_l;i++)
        L[i]=A[p+i];
    for(j=0;j<n_r;j++)
        R[j]=A[q+1+j];

    i=j=0;
    int k=p;

    while(i<n_l && j<n_r)
    {
        if(L[i]<=R[j])
        {
            A[k]=L[i];
            i++;
            k++;
        }
        else
        {
            A[k]=R[j];
            j++;
            k++;
        }
    }

    while (i<n_l)
    {
        A[k]=L[i];
        k++;
        i++;
    }

    while (j<n_r)
    {
        A[k]=R[j];
        k++;
        j++;
    }
}

void Merge_sort(int A[],int p, int r)
{
    if (p>=r)
        return; 
    
    int q=(p+r)/2;
    Merge_sort(A,p,q);
    Merge_sort(A,q+1,r);
    Merge(A,p,q,r);
}

int main()
{
    int arr[]={3,1,2,4};
    Merge_sort(arr,0,3);

    int i;
    for (i=0;i<4;i++)
        cout << arr[i] << ' ';

    return 0;
}

File: Python/CLRS/Merge Sort.py
def Merge_Sort(A: list[int], p: int, r: int):
    if p>=r:
        return
    q = (p + r) // 2
    Merge_Sort(A, p, q)
    Merge_Sort(A, q+1, r)
    Merge(A, p, q, r)

    return A

def Merge(A: list[int], p: int, q: int, r: int):
    n_l = q - p + 1
    n_r = r - q

    L = []
    R = []

    for i in range(n_l):
        L.append(A[p + i])

    for j in range(n_r):
        R.append(A[q + 1 + j])

    i = j = 0
    k = p
    while i < n_l and j < n_r:
        if L[i] <= R[j]:
            A[k] = L[i]
            i += 1
            k += 1
        else:
            A[k] = R[j]
            j += 1
            k += 1

    while i < n_l:
        A[k] = L[i]
        i += 1
        k += 1

    while j < n_r:
        A[k] = R[j]
        j += 1
        k += 1

    return A


print(Merge_Sort([4,3,1,2],0,3))